src/software/ai/hl/stp/play/play.cpp

```diff
@@ -65,54 +77,215 @@ PriorityTacticVector Play::getTactics(const World &world)
+                tactic_vector.push_back(stop_tactics[ii]);
             }
         }
+        num_tactics = tactic_vector.size();
+
+        std::vector<std::unique_ptr<TbotsProto::PrimitiveSet>> primitive_sets;
+
+        for (auto tactic : tactic_vector)
+        {
+            auto motion_constraints =
+                buildMotionConstraintSet(world.gameState(), *goalie_tactic);
+            primitive_sets.emplace_back(getPrimitivesFromTactic(
+                path_planner_factory, world, tactic, motion_constraints));
+            CHECK(primitive_sets.back()->robot_primitives().size() ==
+                  world.friendlyTeam().numRobots())
+                << primitive_sets.back()->robot_primitives().size() << " primitives from "
+                << objectTypeName(*tactic)
+                << " is not equal to the number of robots, which is "
+                << world.friendlyTeam().numRobots();
+        }
+
+        size_t num_rows = robots.size();
+        size_t num_cols = tactic_vector.size();
+
+        // The Matrix constructor will assert if the rows and columns of the matrix are
+        // not >= 1, so we perform that check first and skip over this tactic_vector if
+        // it is empty. This represents the cases where there are either no tactics or no
+        // robots
+        if (num_rows == 0 || num_cols == 0)
+        {
+            continue;
+        }
+
+        // The rows of the matrix are the "workers" (the robots) and the columns are the
+        // "jobs" (the Tactics).
+        Matrix<double> matrix(num_rows, num_cols);
+
+        // Initialize the matrix with the cost of assigning each Robot to each Tactic
+        for (size_t row = 0; row < num_rows; row++)
+        {
+            for (size_t col = 0; col < num_cols; col++)
+            {
+                Robot robot                    = robots.at(row);
+                std::shared_ptr<Tactic> tactic = tactic_vector.at(col);
+                auto primitives = primitive_sets.at(col)->robot_primitives();
+                CHECK(primitives.contains(robot.id()))
+                    << "Couldn't find a primitive for robot id " << robot.id();
+                double robot_cost_for_tactic = primitives.at(robot.id()).cost();
+
+                std::set<RobotCapability> required_capabilities =
+                    tactic->robotCapabilityRequirements();
+                std::set<RobotCapability> robot_capabilities =
+                    robot.getAvailableCapabilities();
+                std::set<RobotCapability> missing_capabilities;
+                std::set_difference(
+                    required_capabilities.begin(), required_capabilities.end(),
+                    robot_capabilities.begin(), robot_capabilities.end(),
+                    std::inserter(missing_capabilities, missing_capabilities.begin()));
+
+                if (missing_capabilities.size() > 0)
+                {
+                    matrix(row, col) = robot_cost_for_tactic * 10.0 + 10.0;
```

Comment:
some magic numbers here :magic_wand: 

would be nice to have a comment explaining why we do this

Reply:
added